include/seqan3/search/detail/search_scheme_algorithm.hpp

```diff
@@ -28,6 +28,112 @@ namespace seqan3::detail
+     * \tparam index_t The type of index; must model seqan3::bi_fm_index_specialisation.
+     * \param[in] cfg The configuration object that guides the search algorithm.
+     * \param[in] index The index used in the algorithm.
+     */
+    search_scheme_algorithm(configuration_t const & cfg, index_t const & index) : policies_t{}...
+    {
+        config = cfg;
+        index_ptr = &index;
+    }
+    //!\}
+
+    /*!\brief Searches a query sequence in a bidirectional index.
+     * \tparam query_t Must model std::ranges::input_range over the index's alphabet.
+     * \param[in] query Query sequence to be searched in the index.
+     *
+     * ### Complexity
+     *
+     * \f$O(|query|^e)\f$ where \f$e\f$ is the total number of maximum errors.
+     */
+    template <typename query_t>
+    auto operator()(query_t && query) noexcept
+    {
+        auto max_error = this->get_search_param(config, query); // see policy_max_error
+
+        // construct internal delegate for collecting hits for later filtering (if necessary)
+        std::vector<typename index_t::cursor_type> internal_hits;
+        auto on_hit_delegate = [&internal_hits] (auto const & it)
+        {
+            internal_hits.push_back(it);
+        };
+
+        // choose mode
+        if constexpr (traits_t::search_best_hits || traits_t::search_all_best_hits || traits_t::search_strata_hits)
+        {
+            auto max_total = max_error.total;
+            max_error.total = 0; // start search with less errors
+            while (internal_hits.empty() && max_error.total <= max_total)
+            {
+                // * If you only want the best hit (traits_t::search_best_hits), you stop after finding the
+                //   first hit, the hit with the least errors (`abort_on_hit` is true).
+                // * If you are in strata mode (traits_t::search_strata_hits), you do the same as with best hits,
+                //   but then do the extra step afterwards (`abort_on_hit` is true).
+                // * If you want all best hits (traits_t::search_all_best_hits), you do not stop after the first
+                //   hit but continue the current search algorithm/max_error pattern (`abort_on_hit` is true).
+                constexpr bool abort_on_hit = !traits_t::search_all_best_hits;
+                search_algo_bi<abort_on_hit>(query, max_error, on_hit_delegate);
+                max_error.total++;
+            }
+            if constexpr (traits_t::search_strata_hits)
+            {
+                if (!internal_hits.empty())
+                {
+                    internal_hits.clear(); // TODO:don't clear when using Optimum Search Schemes with lower error bounds
+                    uint8_t const s = get<search_cfg::mode>(config).value;
+                    max_error.total += s - 1;
+                    search_algo_bi<false>(query, max_error, on_hit_delegate);
+                }
+            }
+        }
+        else // detail::search_mode_all
```

Comment:
In order to make those parts a policy, we would need to transform all policies into CRTP classes because they all need to call the private function `search_algo_bi`/`search_trivial`. I tried to avoid this as I think it's quite complicated.
Can this be part of a later refactoring?

Reply:
I don't think that you need a CRTP base class, because your policy `search_algo_bi` and `search_trivial` could have a single function called `underlying_search`.

I agree with postponing doing this as policies, but what limits you to split this into multiple smaller functions and delegate them in `operator()`?